optimum/gptq/quantizer.py

```diff
@@ -0,0 +1,629 @@
+            raise ValueError
+
+        handle = blocks[0].register_forward_pre_hook(store_input_hook, with_kwargs=True)
+        for data in dataset:
+            for k, v in data.items():
+                # put the data on gpu, we won't put them back to cpu
+                data[k] = v.to(0)
+            try:
+                model(**data)
+            except ValueError:
+                pass
+
+        handle.remove()
+        if not has_device_map:
+            blocks[0].to(device)
+            for module_name in self.module_name_preceding_first_block:
+                module = recurse_getattr(model, module_name)
+                if module is None:
+                    raise ValueError(f"Module {module_name} was not found in model")
+
+        torch.cuda.empty_cache()
+
+        # Step 3: Quantize the blocks
+        quantizers = {}
+        for i, block in enumerate(
+            tqdm(blocks, disable=not verbose, desc=f"Quantizing {self.block_name_to_quantize} blocks ")
+        ):
+            logger.info(f"Start quantizing block {self.block_name_to_quantize} {i + 1}/{len(blocks)}")
+            # move block to cuda if needed
+            # in case we have offload modules, we need to put them on cuda because of GPTQ object
+            if not has_device_map or get_device(block) == torch.device("cpu"):
+                block = block.to(0)
+            layers = get_layers(block)
+            if self.true_sequential:
+                # lazy sequential but works well
+                layers_name_list = [[key] for key in layers.keys()]
+            else:
+                layers_name_list = [list(layers.keys())]
+            logger.info(f"Module to quantize {layers_name_list}")
+            for subset_name_list in tqdm(
+                layers_name_list, disable=not verbose, leave=False, desc="Quantizing layers inside the block"
+            ):
+                subset_layers = {name: layers[name] for name in subset_name_list}
+                gptq = {}
+                handles = []
+                # add hook for each layer in subset_layers
+                for name in subset_layers:
+                    gptq[name] = GPTQ(subset_layers[name])
+                    gptq[name].quantizer.configure(bits=self.bits, sym=self.sym, perchannel=True)
+
+                    def add_batch(name):
+                        def tmp(_, input, output):
+                            gptq[name].add_batch(input[0].data, output.data)
+
+                        return tmp
+
+                    # because it adding a hook will replace the old one.
+                    handles.append(subset_layers[name].register_forward_hook(add_batch(name)))
+                # update Hessian for each layer in subset_layers thanks to the hook
+                for j in range(len(dataset)):
+                    # the args are already on the gpu
+                    # don't need to store the output
+                    block(layer_inputs[j], **layer_input_kwargs[j])
```

Comment:
There's something I don't get here. The `layer_inputs` are the inputs to the first layer in `blocks`. For example, for neox, the inputs to the first `GPTNeoXLayer` from `self.layers =  nn.ModuleList([GPTNeoXLayer(config) for _ in range(config.num_hidden_layers)])`.

Here, we use the inputs of the first layer for the n-th layer?

Reply:
No, we use the output of the n-1 layer as the inputs of the n layers. This is done by getting the outputs and with: 
`layer_inputs, layer_outputs = layer_outputs, []`

Reply:
Thank you, I missed it!